File: frontend/src/lib/features/donations/domain/compute-needs.ts
import type { Donation, DonationCampaign, StockLedger } from '$lib/features/operations';
import { isDonationOutstanding, keyedDonationIds, stockBalance } from '$lib/features/operations';
import { addQty, subQty, qtyGt } from '$lib/utils/qty';

/**
 * Compute remaining needs (remaining = target − donated) and a map of item → campaign.
 *
 * Shared by:
 * - `POST /api/public/v1/donations` (NEED_FULL re-check)
 * - `GET  /api/public/v1/needs`      (needs board aggregation)
 *
 * This avoids three copies of campaign−donated logic drifting apart.
 *
 * @returns `remaining` — Map<item_id, qty_open> summed across campaigns (may be ≤ 0)
 * @returns `itemCampaign` — Map<item_id, campaign_id> (first open campaign that needs the item)
 * @returns `campaignRemaining` — Map<campaign_id, Map<item_id, qty_open>>, the per-campaign
 *   split behind `remaining`. A donation carries exactly one `campaign_id`, so the summed
 *   figure is not by itself bookable — see `pickCampaignForItems`.
 */
export function computeNeeds(
	campaigns: DonationCampaign[],
	donations: Donation[],
	stockLedgers: StockLedger[] = []
): {
	remaining: Map<string, string>;
	itemCampaign: Map<string, string>;
	campaignRemaining: Map<string, Map<string, string>>;
} {
	const remaining = new Map<string, string>();
	const itemCampaign = new Map<string, string>();
	const campaignRemaining = new Map<string, Map<string, string>>();

	// Same terms as the back-office board (deriveNeedAvailability): what the shelter
	// already holds plus what is still owed to it. Deliberately not `calculateReserved` —
	// that one attributes a donation with no campaign_id to every campaign and guesses an
	// item_id from free text, both of which this side dropped on purpose.
	const onHand = stockBalance(stockLedgers);
	const keyed = keyedDonationIds(stockLedgers);

	for (const campaign of campaigns) {
		const covered = new Map(onHand);
		for (const don of donations) {
			if (don.campaign_id !== campaign._id) continue;
			// Terminal without goods ever landing here — `redirected`/`rejected` joined
			// the enum with CR-052 and release their share just like expiry does.
			if (!isDonationOutstanding(don.status) && don.status !== 'received') continue;
			// Received *and* already booked into the ledger: the goods are on the shelf,
			// counted in onHand. Counting them again here closes the need at half.
			if (don.status === 'received' && keyed.has(don._id)) continue;
			for (const it of don.items ?? []) {
				if (!it.item_id) continue;
				covered.set(it.item_id, addQty(covered.get(it.item_id) ?? '0', it.qty));
			}
		}
		for (const need of campaign.needs) {
			// A need staff closed by hand takes no more (T-22 manual force cut-off,
			// CR-052). It still contributes a zero rather than being skipped: callers
			// treat a missing key as "not tracked" and let the booking through, so
			// dropping it here would reopen exactly what the close was meant to stop.
			const rem =
				need.status === 'closed' ? '0' : subQty(need.qty_target, covered.get(need.item_id) ?? '0');
			remaining.set(need.item_id, addQty(remaining.get(need.item_id) ?? '0', rem));
			const perItem = campaignRemaining.get(campaign._id) ?? new Map<string, string>();
			perItem.set(need.item_id, rem);
			campaignRemaining.set(campaign._id, perItem);
			// Bind the item to a campaign still accepting it, so a closed need in one
			// campaign cannot capture a donation another campaign can actually take.
			if (need.status !== 'closed' && !itemCampaign.has(need.item_id)) {
				itemCampaign.set(need.item_id, campaign._id);
			}
		}
	}
	return { remaining, itemCampaign, campaignRemaining };
}

/** What a donation asks for, as far as campaign selection is concerned. */
export type RequestedItem = { item_id?: string | null; qty: string | number };

export type CampaignPick = { ok: true; campaignId: string | null } | { ok: false; itemId: string };

/**
 * Choose the campaign that can actually take this donation.
 *
 * A donation carries exactly one `campaign_id`, but `computeNeeds().remaining` sums an
 * item across every open campaign asking for it. Binding to the first campaign that
 * merely *mentions* the item — the old behaviour — meant a donor could be shown 590 kg
 * of headroom, request 200, and be refused because the campaign they were bound to had
 * only 90 left while another campaign sat on 500 nobody could reach.
 *
 * So pick a campaign whose own remaining covers **every** quota-tracked item in the
 * request. Campaign order decides ties, which keeps the choice deterministic and keeps
 * older campaigns filling first.
 *
 * Returns `campaignId: null` when nothing in the request is quota-tracked (free-text
 * only, or items no open campaign asks for) — those bypass the counter as they always
 * have. Returns `ok: false` with the offending item when no single campaign can serve
 * the request; the summed figure was never bookable in one booking.
 */
export function pickCampaignForItems(
	campaignRemaining: Map<string, Map<string, string>>,
	items: RequestedItem[]
): CampaignPick {
	const perCampaign = [...campaignRemaining.entries()];
	const fits = (perItem: Map<string, string>, item: RequestedItem) =>
		!qtyGt(item.qty, perItem.get(item.item_id as string) ?? '0');

	const tracked = items.filter(
		(item) =>
			!!item.item_id && perCampaign.some(([, perItem]) => perItem.has(item.item_id as string))
	);
	if (tracked.length === 0) return { ok: true, campaignId: null };

	for (const [campaignId, perItem] of perCampaign) {
		if (tracked.every((item) => fits(perItem, item))) return { ok: true, campaignId };
	}

	// Name the item no campaign can serve on its own; if each is individually servable
	// but no campaign covers the whole request, the first tracked item stands for it.
	const offender =
		tracked.find((item) => !perCampaign.some(([, perItem]) => fits(perItem, item))) ?? tracked[0];
	return { ok: false, itemId: offender.item_id as string };
}

File: worker/src/worker/projectors/compute_needs.py
"""Remaining public need per item — the port of the back-office cut-off formula.

T-22 defines one rule: a need closes when **on-hand + reserved ≥ target**. The
back-office board has computed it that way since CR-034 (`deriveNeedAvailability`
takes the stock ledger). This projector, which feeds the *public* board, used to
subtract only donations and never looked at the warehouse — so a shelter holding
540 kg of rice against a 500 kg target still advertised "ด่วน! ขาด 450 กก." while
its own staff screen showed FULL. Donors were being sent goods the shelter had
already closed intake on.

Two formulas for one rule is what allowed that, so this mirrors the TypeScript one
term for term:

    remaining = target − (on_hand + reserved)

where ``on_hand`` sums the stock ledger (issues are negative rows) and ``reserved``
counts donations still owed to the shelter. A ``received`` donation already written
into the ledger is *excluded* from reserved: its quantity is on the shelf now, and
counting it in both places would close the need at half the goods.
"""

from __future__ import annotations

from typing import Any

#: A donation stops owing the shelter anything once it is expired or cancelled. The
#: wider set is kept deliberately — narrowing it here would change what the public board
#: counts, which is a separate decision from wiring the warehouse in.
# Settled: the goods are never arriving at this shelter, so they cannot cover a
# need. "redirected" joins the set with CR-087 (handed to another shelter);
# "rejected" belongs here for the same reason and had been missed.
_SETTLED_STATUSES = frozenset({"expired", "cancelled", "redirected", "rejected"})


def _to_float(value: Any) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0


def _keyed_donation_ids(stock_ledgers: list[dict[str, Any]]) -> set[str]:
    """Donations already booked into the ledger, so their goods count as on-hand."""
    return {
        str(entry["ref_id"])
        for entry in stock_ledgers
        if entry.get("reason") == "donation" and entry.get("ref_id")
    }


def on_hand_by_item(stock_ledgers: list[dict[str, Any]]) -> dict[str, float]:
    balance: dict[str, float] = {}
    for entry in stock_ledgers:
        item_id = entry.get("item_id")
        if not item_id:
            continue
        balance[item_id] = balance.get(item_id, 0.0) + _to_float(entry.get("qty"))
    return balance


def _reserved_by_item(
    donations: list[dict[str, Any]],
    stock_ledgers: list[dict[str, Any]],
    campaign_id: str,
) -> dict[str, float]:
    keyed = _keyed_donation_ids(stock_ledgers)
    reserved: dict[str, float] = {}
    for donation in donations:
        if donation.get("campaign_id") != campaign_id:
            continue
        status = donation.get("status")
        if status in _SETTLED_STATUSES:
            continue
        # Received *and* already booked into the ledger: the goods are on the shelf,
        # counted in on_hand. Counting them again here closes the need at half.
        if status == "received" and str(donation.get("_id")) in keyed:
            continue
        for item in donation.get("items") or []:
            item_id = item.get("item_id")
            if not item_id:
                continue
            reserved[item_id] = reserved.get(item_id, 0.0) + _to_float(item.get("qty"))
    return reserved


def compute_needs(
    campaigns: list[dict[str, Any]],
    donations: list[dict[str, Any]],
    stock_ledgers: list[dict[str, Any]] | None = None,
) -> tuple[dict[str, str], dict[str, str]]:
    """Remaining need per item, summed over campaigns, and the item → campaign map.

    ``stock_ledgers`` defaults to empty so existing callers keep working, but the
    public projector must pass it — without the warehouse the answer is wrong in
    exactly the direction that over-collects.
    """
    ledgers = stock_ledgers or []
    on_hand = on_hand_by_item(ledgers)

    remaining: dict[str, float] = {}
    item_campaign: dict[str, str] = {}

    for campaign in campaigns:
        campaign_id = str(campaign.get("_id", ""))
        reserved = _reserved_by_item(donations, ledgers, campaign_id)

        for need in campaign.get("needs") or []:
            item_id = need.get("item_id")
            if not item_id:
                continue
            # A need staff closed by hand takes no more (T-22 manual force cut-off,
            # CR-052). Contributes a zero rather than being skipped: a missing key reads
            # as "not tracked" downstream and lets the booking through.
            is_closed = need.get("status") == "closed"
            covered = on_hand.get(item_id, 0.0) + reserved.get(item_id, 0.0)
            rem = 0.0 if is_closed else _to_float(need.get("qty_target")) - covered
            remaining[item_id] = remaining.get(item_id, 0.0) + rem
            # Bind the item to a campaign still accepting it.
            if not is_closed and item_id not in item_campaign:
                item_campaign[item_id] = campaign_id

    return (
        {item_id: str(qty) for item_id, qty in remaining.items()},
        item_campaign,
    )


def need_breakdown(
    campaigns: list[dict[str, Any]],
    donations: list[dict[str, Any]],
    stock_ledgers: list[dict[str, Any]] | None = None,
) -> dict[str, dict[str, float]]:
    """The three terms behind ``remaining``, per item, for the public board to show.

    ``compute_needs`` publishes one number — the shortage — which is all the booking
    guard needs. The donor board also wants to say WHY it is short: how much the shelf
    already holds and how much other donors have promised. Those are the same terms this
    module already computes, so they are returned here rather than re-derived in the UI,
    where they had been faked (``target = qty × 2``, ``reserved = 0``).

    Deliberately a sibling of ``compute_needs`` and not a change to it: the shortage rule
    is pinned across three implementations by ``packages/needs-fixtures``, and widening
    its return type would drag the parity harness along for a display concern.

    Item-level aggregate, matching how the board is keyed (``{shelter}:{item_id}``):
    ``qty_target`` sums every open need asking for the item, ``reserved`` sums what is
    still owed across campaigns, and ``on_hand`` is the shelf balance — counted ONCE per
    item even when several campaigns ask for it, the same simplification ``compute_needs``
    makes when it credits the full balance to each campaign.
    """
    ledgers = stock_ledgers or []
    on_hand = on_hand_by_item(ledgers)

    breakdown: dict[str, dict[str, float]] = {}

    def bucket(item_id: str) -> dict[str, float]:
        return breakdown.setdefault(
            item_id,
            {"qty_target": 0.0, "on_hand": on_hand.get(item_id, 0.0), "reserved": 0.0},
        )

    for campaign in campaigns:
        campaign_id = str(campaign.get("_id", ""))
        reserved = _reserved_by_item(donations, ledgers, campaign_id)

        for need in campaign.get("needs") or []:
            item_id = need.get("item_id")
            # A need staff closed by hand takes no more, so it announces no target —
            # same treatment `compute_needs` gives it.
            if not item_id or need.get("status") == "closed":
                continue
            entry = bucket(str(item_id))
            entry["qty_target"] += _to_float(need.get("qty_target"))
            entry["reserved"] += reserved.get(item_id, 0.0)

    return breakdown
